File: functions/invitations/resend_invitation.py
from datetime import datetime, timedelta, timezone

from firebase_admin import firestore
from flask import abort
from models.constants import Collections, InvitationFields, Status
from models.data_models import Invitation
from utils.logging_utils import get_logger


def resend_invitation(request, invitation_id) -> Invitation:
    """
    Resends an invitation by resetting its created_at time and updating the expires_at time.

    Args:
        request: The Flask request object containing:
                - user_id: The authenticated user's ID (attached by authentication middleware)
        invitation_id: The ID of the invitation to resend

    Returns:
        The updated Invitation object with refreshed timestamps

    Raises:
        403: You can only resend your own invitations
        404: Invitation not found
    """
    logger = get_logger(__name__)
    logger.info(f"User {request.user_id} resending invitation {invitation_id}")

    current_user_id = request.user_id

    # Initialize Firestore client
    db = firestore.client()

    # Get the invitation document
    invitation_ref = db.collection(Collections.INVITATIONS).document(invitation_id)
    invitation_doc = invitation_ref.get()

    # Check if the invitation exists
    if not invitation_doc.exists:
        logger.warning(f"Invitation {invitation_id} not found")
        abort(404, description="Invitation not found")

    invitation_data = invitation_doc.to_dict()

    # Check if the current user is the sender of the invitation
    sender_id = invitation_data.get(InvitationFields.SENDER_ID)
    if sender_id != current_user_id:
        logger.warning(
            f"User {current_user_id} is not the sender of invitation {invitation_id}"
        )
        abort(403, description="You can only resend your own invitations")

    # Set new timestamps
    current_time = datetime.now(timezone.utc)
    expires_at = current_time + timedelta(days=1)

    # Update the invitation with new timestamps
    invitation_ref.update(
        {
            InvitationFields.CREATED_AT: current_time,
            InvitationFields.EXPIRES_AT: expires_at,
            InvitationFields.STATUS: Status.PENDING,
        }
    )

    logger.info(f"User {current_user_id} resent invitation {invitation_id}")

    # Return the updated invitation
    return Invitation(
        invitation_id=invitation_id,
        created_at=current_time.isoformat(),
        expires_at=expires_at.isoformat(),
        sender_id=current_user_id,
        status=Status.PENDING,
        username=invitation_data.get(InvitationFields.USERNAME, ""),
        name=invitation_data.get(InvitationFields.NAME, ""),
        avatar=invitation_data.get(InvitationFields.AVATAR, ""),
    )

File: functions/src/invitations/resend-invitation.ts
import { Request } from 'express';
import { DocumentData, getFirestore, Timestamp, UpdateData, } from 'firebase-admin/firestore';
import { ApiResponse, EventName, InviteEventParams, } from '../models/analytics-events.js';
import { Collections, InvitationFields, Status } from '../models/constants.js';
import { Invitation } from '../models/data-models.js';
import { BadRequestError, ForbiddenError, NotFoundError, } from '../utils/errors.js';
import { hasReachedCombinedLimit } from '../utils/friendship-utils.js';
import { calculateExpirationTime } from '../utils/invitation-utils.js';
import { getLogger } from '../utils/logging-utils.js';
import { formatTimestamp } from '../utils/timestamp-utils.js';
import { hasLimitOverride } from '../utils/profile-utils.js';

import { fileURLToPath } from 'url';
import path from 'path';

const __filename = fileURLToPath(import.meta.url);
const logger = getLogger(path.basename(__filename));

/**
 * Resends an invitation by resetting its created_at time and updating the expires_at time.
 *
 * Validates that:
 * 1. The user hasn't reached the combined limit of friends and active invitations (5)
 *
 * @param req - The Express request object containing:
 *              - userId: The authenticated user's ID (attached by authentication middleware)
 *              - params: The route parameters containing:
 *                - invitation_id: The ID of the invitation to resend
 *
 * @returns An ApiResponse containing the updated invitation and analytics
 *
 * @throws {400} User has reached the maximum number of friends and active invitations
 * @throws {403} You can only resend your own invitations
 * @throws {404} Invitation not found
 */
export const resendInvitation = async (
  req: Request,
): Promise<ApiResponse<Invitation>> => {
  const currentUserId = req.userId;
  const invitationId = req.params.invitation_id;
  logger.info(`User ${currentUserId} resending invitation ${invitationId}`);

  // Initialize Firestore client
  const db = getFirestore();

  if (!invitationId) {
    throw new BadRequestError('Invitation ID is required');
  }

  // Get the invitation document
  const invitationRef = db
    .collection(Collections.INVITATIONS)
    .doc(invitationId);
  const invitationDoc = await invitationRef.get();

  // Check if the invitation exists
  if (!invitationDoc.exists) {
    logger.warn(`Invitation ${invitationId} not found`);
    throw new NotFoundError('Invitation not found');
  }

  const invitationData = invitationDoc.data();

  // Check if the current user is the sender of the invitation
  const senderId = invitationData?.[InvitationFields.SENDER_ID];
  if (senderId !== currentUserId) {
    logger.warn(
      `User ${currentUserId} is not the sender of invitation ${invitationId}`,
    );
    throw new ForbiddenError('You can only resend your own invitations');
  }

  // Check combined limit (excluding the current invitation)
  const { friendCount, activeInvitationCount, hasReachedLimit } =
    await hasReachedCombinedLimit(currentUserId, invitationId);
  if (hasReachedLimit) {
    const override = await hasLimitOverride(currentUserId);
    if (!override) {
      logger.warn(
        `User ${currentUserId} has reached the maximum number of friends and active invitations`,
      );
      throw new BadRequestError(
        'You have reached the maximum number of friends and active invitations',
      );
    }
  }

  // Set new timestamps
  const currentTime = Timestamp.now();
  const expiresAt = calculateExpirationTime();

  // Update the invitation with new timestamps
  const updatePayload: UpdateData<DocumentData> = {
    [InvitationFields.CREATED_AT]: currentTime,
    [InvitationFields.EXPIRES_AT]: expiresAt,
    [InvitationFields.STATUS]: Status.PENDING,
  };
  await invitationRef.update(updatePayload);

  logger.info(`User ${currentUserId} resent invitation ${invitationId}`);

  // Return the updated invitation
  const invitation: Invitation = {
    invitation_id: invitationId,
    created_at: formatTimestamp(currentTime),
    expires_at: formatTimestamp(expiresAt),
    sender_id: currentUserId,
    status: Status.PENDING,
    username: invitationData?.[InvitationFields.USERNAME] || '',
    name: invitationData?.[InvitationFields.NAME] || '',
    avatar: invitationData?.[InvitationFields.AVATAR] || '',
    receiver_name: invitationData?.[InvitationFields.RECEIVER_NAME] || '',
  };

  // Create analytics event
  const event: InviteEventParams = {
    friend_count: friendCount,
    invitation_count: activeInvitationCount + 1, // Add 1 for the resent invitation
  };

  return {
    data: invitation,
    status: 200,
    analytics: {
      event: EventName.INVITE_RESENT,
      userId: currentUserId,
      params: event,
    },
  };
};
